File: cp/src/neetcode150/stack/leetcode20.rs
#![allow(dead_code)]
struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = Vec::new();
        for ch in s.chars() {
            match ch {
                '(' => stack.push(')'),
                '{' => stack.push('}'),
                '[' => stack.push(']'),
                n => {
                    if stack.is_empty() {
                        return false;
                    }

                    if let Some(&top) = stack.last() {
                        if top != n {
                            return false;
                        }
                    }
                    stack.pop();
                }
            }
        }

        stack.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    pub fn is_valid_test_case1() {
        let s = String::from("()[]{}");
        assert_eq!(Solution::is_valid(s), true);

        let s = String::from("(]");
        assert_eq!(Solution::is_valid(s), false);
    }
}

File: js_algorithms/neetcode150/stack/leetcode20.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function (s) {
  const stack = [];

  for (const ch of s) {
    if (ch === '(') {
      stack.push(')');
    } else if (ch === '[') {
      stack.push(']');
    } else if (ch === '{') {
      stack.push('}');
    } else if (!stack.length || stack.pop() !== ch) {
      return false;
    }
  }

  return !stack.length;
};

module.exports = isValid;
